app/models/spree/payment.rb

```diff
@@ -160,6 +160,14 @@ def validate_source
       errors.blank?
     end
 
+    def valid_source?
+      if source.is_a?(Spree::Payment)
+        true
+      else # top-most source
```

Comment:
this is very hard for me to understand even if I read the commit message and if know the context inside out...
what if we extract this little method to a little service and add a nice comment to explain?

Reply:
I think I can manage with a good method doc. This won't be pretty as we are approaching callbacks nastiness but bear with me.